### ISSUE TITLE
How to handle a reference genome ?

### ISSUE THREAD

**Issue by GDelevoye** [deeptools/pyGenomeTracks#352]
Hello,

I am new to the pyGenomeTracks but very surprised, as it doesn't seem like I can display any reference genome / fasta file with it ? I really thought this would be first thing that would be implemented in a genome browser

Do you know if there is any fork somewhere, pending PR or practical workaround you could think of to display the genome sequence from a fasta file ?

**Comment by lldelisle** [comment]
Hi,
If I understand correctly you would like to display the sequence?
You can write a PR for sure, this would be a new track type. See the documentation [here](https://pygenometracks.readthedocs.io/en/latest/content/adding-new-tracks.html).
I wrote something for the MAF which can be useful for you but is still in development:
https://github.com/deeptools/pyGenomeTracks/blob/b95f8fc30b86798c6c97879c0c0a6199f151b619/pygenometracks/tracks/MafTrack.py#L12-L16
https://github.com/deeptools/pyGenomeTracks/blob/b95f8fc30b86798c6c97879c0c0a6199f151b619/pygenometracks/tracks/MafTrack.py#L263-L282

The output looks like:

![this](https://github.com/deeptools/pyGenomeTracks/raw/maf/pygenometracks/tests/test_data/master_first_maf_seq.png)

or

![that](https://github.com/deeptools/pyGenomeTracks/raw/maf/pygenometracks/tests/test_data/master_first_maf_seq_zoom.png)

**Comment by GDelevoye** [comment]
Hi @lldelisle , thank you very much for your help

Starting with your code I came to this :

```python
from . GenomeTrack import GenomeTrack
from .. utilities import get_optimal_fontsize
import numpy as np
import os

# Color code for seqs
seq_color = {'A': 'red',
             'T': 'green',
             'G': 'blue',
             'C': 'black'}


def load_fasta(fastafile):
    """Returns a python dict { id : sequence } for the given .fasta file"""
    with open(os.path.realpath(fastafile), 'r') as filin:
        fasta = filin.read()
        fasta = fasta.split('>')[1:]
        outputdict = {x.split('\n')[0].strip(): "".join(x.split('\n')[1:]) for x in fasta}
    return outputdict


class FaTrack(GenomeTrack):
    SUPPORTED_ENDINGS = ['.fa',".fasta"]
    TRACK_TYPE = 'fasta'
    OPTIONS_TXT = GenomeTrack.OPTIONS_TXT

    DEFAULTS_PROPERTIES = {'line_width': 0.5,"height":1,"title":"Reference"}
    NECESSARY_PROPERTIES = ['file']
    SYNONYMOUS_PROPERTIES = {}
    POSSIBLE_PROPERTIES = {}
    BOOLEAN_PROPERTIES = []
    STRING_PROPERTIES = ["file","title"]
    FLOAT_PROPERTIES = {'line_width': [0, np.inf],
                        'height': [0, np.inf]}
    INTEGER_PROPERTIES = {}

    def __init__(self, *args, **kwarg):
        super(FaTrack, self).__init__(*args, **kwarg)
        self.ref = self.properties['file']
        self.seq = load_fasta(self.ref)

    def plot(self, ax, chrom_region, start_region, end_region):
        plotting_figure_width = ax.get_window_extent().transformed(ax.get_figure().dpi_scale_trans.inverted()).width

        # The first constrain on the fontsize is the width
        ideal_fontsize = 1.4 * get_optimal_fontsize(plotting_figure_width,
                                              start_region, end_region)
        # The other constraint is the height
        # 1 point = 1/72 inch = height of character
        cm = 2.54
        max_fontsize = self.properties["height"] * 72 / cm

        # Let's take the biggest font possible with these constraints so that the figure is as readable as possible
        fontsize = min(ideal_fontsize,max_fontsize)

        seq_overlap = self.seq[chrom_region][start_region:end_region+1]
        for i in range(len(seq_overlap)):
            ax.text(i+start_region, 0.5, seq_overlap[i],
                    color=seq_color[seq_overlap[i]], verticalalignment='center',
                    horizontalalignment='center',fontsize=fontsize)

```
I tried to keep it as simple as possible, it seems to work

The font can overflow at left and right of the track for big heights / small number of nucleotides, , sometimes over-riding the name of the track. 

[toobig.pdf](https://github.com/deeptools/pyGenomeTracks/files/6847329/toobig.pdf)

Actually I think the only way to avoid it while still keep the nucleotides aligned with their position would be by adding a left and right margin for each track, which I think is beyond of the future and would request larger rework of the code (?)

It gives, however, to satisfying results with reasonable and sizes closer to real-world ones : 

[test.pdf](https://github.com/deeptools/pyGenomeTracks/files/6847354/test.pdf)

The only thing I couldn't get rid off is the vertical bar at the left. I guess it is because my track is considered by default as a quantitative track. Do you know how I could remove it ?

**Comment by lldelisle** [comment]
Hi,
Please write a PR, this is a good enhancement. I would still need to modify some stuff but it looks very good.
For me a way to prevent writting into margin is that if the fontsize is too big we don't display the first and last letter what do you think?

**Comment by GDelevoye** [comment]
I am not sure I will have the time to do it, but I will try to write a clean PR in the comming days

**Comment by lldelisle** [comment]
I wrote it including your commits... I hope you don't mind.

**Comment by GDelevoye** [comment]
#360 is a very cool PR, thank you ! 
I would not have done anything close to this level even if I had the time for it :)

I'm a bit concerned about the fact of removing the last nucleotide (even if, pragmatically speaking, that's 100% understandable). As a user, if I want to plot the 0th nucleotide to the 10th nucleotide, I'd expect the 50th nucleotide to be plotted toot.

What about using the reverse approach: plot all the tracks up to N+1 except the text tracks ? Idk, that's a real dilemma for me here

Another pragmatic option could be to just give the user a warning if the right margin is not sufficient

**Comment by lldelisle** [comment]
For me if you want to plot chr1:0-10, I take it as BED = 0-based half-open.
That means that you want to see the first 10 nucleotides. I checked on UCSC and this is how they display their sequences.
I know IGV has another approach which is like in GTF = 1-based closed.
This means that you need to say chr1:1-10 to see the same and then the bases are displayed at ticks not in between ticks...
I never heard about a 0-based closed approach.

**Comment by GDelevoye** [comment]
You make a good point with the UCSC standard here. Plus it solves the right-margin issue and fits with the natural way of indexing in python... I can only agree with you and withdraw my previous remark

**Comment by lldelisle** [comment]
Solved in version 3.7
